pysqldb3/geopackage.py

```diff
@@ -521,13 +527,154 @@ def read_gpkg(self, dbo = None, schema = None, table = None, port = 5432, srid =
 
             if not private:
                 try:
-                    dbo.query('grant select on {s}."{t}" to public;'.format(
-                        s=schema, t=output_name),
+                    dbo.query(f'grant select on {schema}."{table}" to public;',
                         timeme=False, internal=True, strict=True)
                 except:
                     pass
             self.rename_geom(dbo, schema, table)
 
+        else:
+
+            ####### BULK UPLOAD ########
+        
+            count_cmd = COUNT_GPKG_LAYERS.format(full_path = full_path) 
+
+            try:
+                ogr_response = subprocess.check_output(shlex.split(count_cmd.replace('\n', ' ')), stderr=subprocess.STDOUT)
+                tables_in_gpkg = re.findall(r"\\n\d+:\s(.*?)(?=\\r|\s\(.*\))", str(ogr_response))
+                # only allows tables names with underscores, numbers, and letters as the first character
+                # excludes any dtype description that's also returned by the command line
+            
+            except subprocess.CalledProcessError as e:
+                print("Ogr2ogr Output:\n", e.output)
+                print('Ogr2ogr command failed. The Geopackage was not read in.')
+                raise subprocess.CalledProcessError(cmd=print_cmd_string([dbo.password], count_cmd), returncode=1)
+            
+
+            # create a list of cleaned gpkg table names
+            gpkg_tbl_names = {} # create empty dictionary
+            for t_i_g in tables_in_gpkg:
+                insert_val = re.sub(r'[^A-Za-z0-9_]+', r'', t_i_g)
+                gpkg_tbl_names[insert_val] = t_i_g # add the cleaned name
+
+            # assert that the new cleaned names are unique. if not, we won't get the same dimensions
+            assert len(gpkg_tbl_names) == len(tables_in_gpkg), "Clean the geopackage table names so they can be uploaded as tables (by removing special characters other than _) and make sure they are unique."
+
+            # loop through the tables     
+            for output_name, input_name in gpkg_tbl_names.items():
+            
+                if dbo.table_exists(table = output_name, schema = schema):
+
+                    print(f'Deleting existing table {schema}.{output_name}')
+                    dbo.drop_table(schema=schema, table=output_name)
+
+                if dbo.type == 'PG':
```

Comment:
this logic was used earlier, can you pull it out into a function that gets the list of table names so the logic only exists once

Reply:
I will just put all of it into a bulk upload function